File: internal/memory/context_attach.go
package memory

import (
	"errors"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/stop1love1/claude-bridge/internal/symbol"
)

// Reference-attach caps. Per-file size mirrors PinnedPerFileCapBytes
// because the prompt rendering treats both kinds of attachment the
// same way; the count cap is intentionally smaller than PinnedMaxFiles
// because reference picks are speculative — quality > quantity.
const (
	ReferencePerFileCapBytes = 4 * 1024
	ReferenceMaxFiles        = 3
	// ReferenceMinScore is the minimum keyword overlap a candidate
	// file needs to earn an attachment slot. Below 2 we'd be injecting
	// noise — a single accidental token match is not signal.
	ReferenceMinScore = 2
)

// stopwords skipped during task-body tokenization. Trimmed to the
// most common spawn-prompt offenders — same list as libs/contextAttach.ts.
// A map gives O(1) membership without pulling in a set type.
var stopwords = map[string]struct{}{
	"the": {}, "and": {}, "a": {}, "an": {}, "of": {}, "to": {},
	"for": {}, "in": {}, "on": {}, "at": {}, "is": {}, "are": {},
	"be": {}, "by": {}, "as": {}, "or": {}, "with": {}, "this": {},
	"that": {}, "add": {}, "fix": {}, "update": {}, "create": {},
	"make": {}, "build": {}, "use": {}, "new": {}, "old": {},
	"from": {}, "into": {}, "out": {}, "do": {}, "did": {}, "done": {},
	"task": {}, "todo": {}, "should": {}, "would": {}, "need": {},
	"needs": {}, "want": {}, "wants": {}, "please": {}, "review": {},
	"implement": {}, "function": {}, "feature": {},
}

// tokenSplitRE matches every run of non-alphanumeric characters — the
// inverse of the tokenizer's "keep-set". Compiled once because the
// hot path is the spawn flow and we don't want to re-parse the regex
// on every prompt build.
var tokenSplitRE = regexp.MustCompile(`[^a-z0-9]+`)

// numericTokenRE drops pure-numeric tokens (years, line numbers,
// version digits) so a task body referencing "2025" or "v3" doesn't
// match every file with those digits in its path.
var numericTokenRE = regexp.MustCompile(`^\d+$`)

// ReferenceFile is one auto-attached file picked by the keyword
// heuristic. Rel is always posix-style so the rendered prompt is
// platform-stable.
type ReferenceFile struct {
	Rel       string
	Content   string
	Truncated bool
	// Score is the summed keyword-overlap score that earned the file
	// its slot. Surfaced in the prompt's "## Reference files" intro
	// so the agent understands why a file was attached.
	Score int
}

// AttachOptions bundles the inputs to AttachReferences. ExcludePaths
// is typically the operator's pinned-file list — we skip those so the
// reference slot doesn't waste space duplicating already-pinned content.
type AttachOptions struct {
	AppPath      string
	ExcludePaths []string
}

// Tokenize lowercases, splits on non-alphanumeric runs, and filters
// out short / stopword / pure-numeric tokens. Returns deduplicated,
// stable-ordered tokens (insertion order on first sight) so two calls
// with the same input produce the same slice — the scoring side relies
// on that determinism only for test reproducibility.
//
// Exported so the recent-direction picker can reuse the exact same
// tokenization as the reference attacher (matches libs/contextAttach.ts
// exporting `tokenize` for `recentDirection.ts` to import).
func Tokenize(text string) []string {
	if text == "" {
		return nil
	}
	lower := strings.ToLower(text)
	pieces := tokenSplitRE.Split(lower, -1)
	seen := make(map[string]struct{}, len(pieces))
	out := make([]string, 0, len(pieces))
	for _, raw := range pieces {
		t := strings.TrimSpace(raw)
		if t == "" || len(t) < 3 {
			continue
		}
		if _, skip := stopwords[t]; skip {
			continue
		}
		if numericTokenRE.MatchString(t) {
			continue
		}
		if _, dup := seen[t]; dup {
			continue
		}
		seen[t] = struct{}{}
		out = append(out, t)
	}
	return out
}

// ScoreSymbol counts substring matches of taskTokens inside the
// symbol's `file + " " + name` haystack. Substring match (not whole
// word) so `formInput` is matched by `form` and `useFormState` is
// matched by `form` — that's how the agent's loose "this is about
// forms" intent maps onto the index.
func ScoreSymbol(s symbol.SymbolEntry, taskTokens []string) int {
	if len(taskTokens) == 0 {
		return 0
	}
	haystack := strings.ToLower(s.File + " " + s.Name)
	score := 0
	for _, tok := range taskTokens {
		if strings.Contains(haystack, tok) {
			score++
		}
	}
	return score
}

// CandidateFile is one file in the scored-and-ranked result of
// PickCandidateFiles. Score is the sum of every symbol in this file
// scored against the task tokens.
type CandidateFile struct {
	File  string
	Score int
}

// PickCandidateFiles aggregates per-symbol scores into per-file scores,
// drops files below ReferenceMinScore, and returns the survivors
// sorted by score descending (file path ascending breaks ties so the
// output is deterministic).
//
// Exported so recent-direction can pick a "touched dir" using the same
// file-ranking logic — keeps the heuristic in one place.
func PickCandidateFiles(symbols []symbol.SymbolEntry, taskTokens []string) []CandidateFile {
	if len(taskTokens) == 0 {
		return nil
	}
	fileScores := make(map[string]int)
	for _, s := range symbols {
		inc := ScoreSymbol(s, taskTokens)
		if inc == 0 {
			continue
		}
		fileScores[s.File] += inc
	}
	out := make([]CandidateFile, 0, len(fileScores))
	for file, score := range fileScores {
		if score < ReferenceMinScore {
			continue
		}
		out = append(out, CandidateFile{File: file, Score: score})
	}
	sort.Slice(out, func(i, j int) bool {
		if out[i].Score != out[j].Score {
			return out[i].Score > out[j].Score
		}
		return out[i].File < out[j].File
	})
	return out
}

// AttachReferences picks up to ReferenceMaxFiles files from index by
// keyword overlap with taskBody, reads each (capped + truncation flag),
// and returns them ranked by score. Returns nil when the index is
// empty, the task body produces no useful tokens, no candidate scored
// above the threshold, or every candidate failed to read.
//
// The signature matches the per-task spawn shape: one task body, one
// shared symbol index per app, attach-time options. Errors at the
// per-file level are swallowed (we just skip that file) — the prompt
// builder doesn't need to know that a candidate file vanished between
// indexing and reading; it just gets fewer references this turn.
func AttachReferences(taskBody string, index symbol.SymbolIndex, opts AttachOptions) []ReferenceFile {
	if len(index.Symbols) == 0 {
		return nil
	}
	tokens := Tokenize(taskBody)
	if len(tokens) == 0 {
		return nil
	}
	if opts.AppPath == "" || !filepath.IsAbs(opts.AppPath) {
		// Without an absolute app root we can't safely resolve any
		// candidate path — bail rather than guessing against cwd.
		return nil
	}

	exclude := make(map[string]struct{}, len(opts.ExcludePaths))
	for _, p := range opts.ExcludePaths {
		exclude[filepath.ToSlash(p)] = struct{}{}
	}

	candidates := PickCandidateFiles(index.Symbols, tokens)
	out := make([]ReferenceFile, 0, ReferenceMaxFiles)
	for _, c := range candidates {
		if len(out) >= ReferenceMaxFiles {
			break
		}
		norm := filepath.ToSlash(c.File)
		if _, skip := exclude[norm]; skip {
			continue
		}
		abs, ok := resolveSafely(opts.AppPath, c.File)
		if !ok {
			continue
		}
		content, truncated, ok := readReferenceCapped(abs)
		if !ok {
			continue
		}
		out = append(out, ReferenceFile{
			Rel:       norm,
			Content:   content,
			Truncated: truncated,
			Score:     c.Score,
		})
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

// readReferenceCapped is the reference-side twin of pinned.go's
// readCapped, kept separate so a future cap divergence between the two
// flows doesn't force one knob for both. Same N+1 buffer trick so a
// file exactly at the cap reports truncated=false.
func readReferenceCapped(absPath string) (string, bool, bool) {
	f, err := os.Open(absPath)
	if err != nil {
		return "", false, false
	}
	defer f.Close()
	buf := make([]byte, ReferencePerFileCapBytes+1)
	n, err := io.ReadFull(f, buf)
	if err != nil && !errors.Is(err, io.EOF) && !errors.Is(err, io.ErrUnexpectedEOF) {
		return "", false, false
	}
	truncated := n > ReferencePerFileCapBytes
	if truncated {
		n = ReferencePerFileCapBytes
	}
	return string(buf[:n]), truncated, true
}

File: libs/contextAttach.ts
import { existsSync, readFileSync } from "node:fs";
import { isAbsolute, relative, resolve } from "node:path";
import type { SymbolIndex, SymbolEntry } from "./symbolIndex";

const PER_FILE_CAP_BYTES = 4 * 1024;
const MAX_REFERENCES = 3;
const MIN_SCORE = 2;
const STOPWORDS = new Set([
  "the", "and", "a", "an", "of", "to", "for", "in", "on", "at",
  "is", "are", "be", "by", "as", "or", "with", "this", "that",
  "add", "fix", "update", "create", "make", "build", "use",
  "new", "old", "from", "into", "out", "do", "did", "done",
  "task", "todo", "should", "would", "need", "needs", "want",
  "wants", "please", "review", "implement", "function", "feature",
]);

export interface ReferenceFile {
  rel: string;
  content: string;
  truncated: boolean;
  score: number;
}

export function tokenize(text: string): string[] {
  const out = new Set<string>();
  for (const raw of text.toLowerCase().split(/[^a-z0-9]+/g)) {
    const t = raw.trim();
    if (!t || t.length < 3) continue;
    if (STOPWORDS.has(t)) continue;
    if (/^\d+$/.test(t)) continue;
    out.add(t);
  }
  return [...out];
}

function resolveSafely(appPath: string, rel: string): string | null {
  if (!rel || isAbsolute(rel)) return null;
  const abs = resolve(appPath, rel);
  const within = relative(appPath, abs);
  if (within.startsWith("..") || isAbsolute(within)) return null;
  return abs;
}

export function scoreSymbol(symbol: SymbolEntry, taskTokens: string[]): number {
  const haystack = (symbol.file + " " + symbol.name).toLowerCase();
  let score = 0;
  for (const tok of taskTokens) {
    if (haystack.includes(tok)) score += 1;
  }
  return score;
}

interface CandidateFile {
  file: string;
  score: number;
}

export function pickCandidateFiles(
  symbols: SymbolEntry[],
  taskTokens: string[],
): CandidateFile[] {
  if (taskTokens.length === 0) return [];
  const fileScores = new Map<string, number>();
  for (const s of symbols) {
    const inc = scoreSymbol(s, taskTokens);
    if (inc === 0) continue;
    fileScores.set(s.file, (fileScores.get(s.file) ?? 0) + inc);
  }
  const out: CandidateFile[] = [];
  for (const [file, score] of fileScores) {
    if (score < MIN_SCORE) continue;
    out.push({ file, score });
  }
  out.sort((a, b) => b.score - a.score || a.file.localeCompare(b.file));
  return out;
}

function readCapped(absPath: string): { content: string; truncated: boolean } | null {
  try {
    const buf = readFileSync(absPath);
    const truncated = buf.byteLength > PER_FILE_CAP_BYTES;
    const content = buf.subarray(0, PER_FILE_CAP_BYTES).toString("utf8");
    return { content, truncated };
  } catch {
    return null;
  }
}

export interface AttachReferencesOptions {
  appPath: string;
  taskBody: string;
  symbolIndex: SymbolIndex | null;
  excludePaths?: string[];
}

export function attachReferences(opts: AttachReferencesOptions): ReferenceFile[] {
  const { appPath, taskBody, symbolIndex } = opts;
  if (!symbolIndex || symbolIndex.symbols.length === 0) return [];
  const tokens = tokenize(taskBody ?? "");
  if (tokens.length === 0) return [];

  const exclude = new Set((opts.excludePaths ?? []).map((p) => p.replace(/\\/g, "/")));
  const candidates = pickCandidateFiles(symbolIndex.symbols, tokens);

  const out: ReferenceFile[] = [];
  for (const c of candidates) {
    if (out.length >= MAX_REFERENCES) break;
    const norm = c.file.replace(/\\/g, "/");
    if (exclude.has(norm)) continue;
    const abs = resolveSafely(appPath, c.file);
    if (!abs || !existsSync(abs)) continue;
    const read = readCapped(abs);
    if (!read) continue;
    out.push({
      rel: norm,
      content: read.content,
      truncated: read.truncated,
      score: c.score,
    });
  }
  return out;
}

export const __test = {
  tokenize,
  scoreSymbol,
  pickCandidateFiles,
  resolveSafely,
  PER_FILE_CAP_BYTES,
  MAX_REFERENCES,
  MIN_SCORE,
};
